crypto/cipher_extra/e_chacha20poly1305.c

```diff
@@ -346,3 +382,320 @@ const EVP_AEAD *EVP_aead_chacha20_poly1305(void) {
+                                     const uint8_t *inp, size_t len)
+{
+  CIPHER_CHACHA_POLY_CTX  *cipher_ctx = CCP_CTX(ctx);
+  CIPHER_CHACHA_KEY *key = CC_KEY(cipher_ctx);
+  uint32_t n, rem, ctr32;
+
+  // Complete any partial block
+  n = key->partial_len;
+  if (n) {
+    // Compute the cipher using our partially used key and any new input up
+    // to the next block
+    while (len && n < CHACHA_BLOCK_LEN) {
+      // Compute 1-byte of output by xor'ing 1-byte of input with the
+      // corresponding key byte and increment it all for the next byte.
+      *out++ = *inp++ ^ key->buf[n++];
+      len--;
+    }
+    key->partial_len = n;
+
+    // If we consumed all the input, we're done
+    if (len == 0) {
+      return 1;
+    }
+
+    // If we completed a block, increment the counter
+    if (n == CHACHA_BLOCK_LEN) {
+      key->partial_len = 0;
+      key->counter_nonce[0]++;
+    }
+  }
+
+  // Truncate down to the last complete block prior to the bulk cipher
+  rem = (uint32_t)(len % CHACHA_BLOCK_LEN);
+  len -= rem;
+  ctr32 = key->counter_nonce[0];
+  while (len >= CHACHA_BLOCK_LEN) {
+    size_t blocks = len / CHACHA_BLOCK_LEN;
+    // 1<<28 is just a not-so-small yet not-so-large number... Below
+    // condition is practically never met, but it has to be checked for code
+    // correctness.
+    if (sizeof(size_t) > sizeof(uint32_t) && blocks > (1U<<28)) {
+      blocks = (1U << 28);
+    }
+
+    // As ChaCha20_ctr32 operates on 32-bit counter, caller has to handle
+    // overflow. 'if' below detects the overflow, which is then handled by
+    // limiting the amount of blocks to the exact overflow point...
+    ctr32 += (uint32_t) blocks;
+    if (ctr32 < blocks) {
+      blocks -= ctr32;
+      ctr32 = 0;
+    }
+    blocks *= CHACHA_BLOCK_LEN;
+    CRYPTO_chacha_20(out, inp, blocks, (uint8_t *) key->key,
+                     (uint8_t *) cipher_ctx->iv, key->counter_nonce[0]);
+    len -= blocks;
+    inp += blocks;
+    out += blocks;
+
+    key->counter_nonce[0] = ctr32;
+  }
+
+  // Start the next block if we have any leftover input
+  if (rem) {
+    memset(key->buf, 0, sizeof(key->buf));
+    // Obtain the current key and store it in the context
+    CRYPTO_chacha_20(key->buf, key->buf, CHACHA_BLOCK_LEN, (uint8_t *) key->key,
+                     (uint8_t *) cipher_ctx->iv, key->counter_nonce[0]);
+    for (n = 0; n < rem; n++) {
+      out[n] = inp[n] ^ key->buf[n];
+    }
+    key->partial_len = rem;
+  }
+
+  return 1;
+}
+
+static int cipher_chacha20_poly1305_do_cipher(
+    EVP_CIPHER_CTX *ctx, unsigned char *out, const unsigned char *inp,
+    size_t inl) {
+  CIPHER_CHACHA_POLY_CTX *cipher_ctx = CCP_CTX(ctx);
+  poly1305_state *poly_ctx = POLY_CTX(cipher_ctx);
+  size_t remainder;
+
+  if (!cipher_ctx->poly_initialized) {
+    // Obtain the poly1305 key by computing the 0th chacha20 key
+    const uint8_t *chacha_key = (const uint8_t *) cipher_ctx->key.key;
```

Comment:
This will also hit an endian issue.

Reply:
Good catch. I can fix this, but it's not gonna be pretty.